### PR TITLE
Webforms compliance

### LINKED ISSUE #29: Compliance with standard web form elements
Currently when you define a selectivity input out of standard HTML's:
* `select`, this element gets replaced completely with selectivity input
* `input`, the logic works, but everything is appended into the `input` tag, creating something like `<input>...</input>`

What about support for standard web forms? If I try to turn `select` into selectivity input, it will end up with removing the select field completely from the form. That shouldn't be the case.

The similar situation is `multiple: true` defined on the `input`. The logic seems to operate, but nothing happens, and field is removed from the form as well.

`Select2` has this not-really-so-great approach of hiding those standard elements and injecting/updating values within once selection is made, but that's actually pretty awesome as it allows users to prettify the select boxes and not modify anything else about the code for it to work as it should (and did).

### PR DESCRIPTION
**Pull request by msx2**
@arendjr This pretty much resolves #29. Changes apply only for selectivity inputs that were initialized from traditional `<select>` boxes with standard `name` attribute set.

By the way, I've also changed `name` to `data-name` in `selectivity-traditional.js`. It was not used anywhere, and now we are able to search using plain `$('[name="xxx"]')` and expect `<select>` elements in result.

### CODE REVIEW

**Comment by arendjr** [comment]
I really like the idea of this patch. Great work! It definitely shows this is more feasible than I would've feared :)

My major concern right now is that most of the logic of keeping the `<select>` element in sync with the selection is implemented in the `single` and `multiple` modules, whereas I would like to see the actual logic implemented in the `traditional` module. I think this would separate concerns better, and will make sure users who are not using the traditional module won't carry its overhead in their bundles.

This would probably require some extra hooks to be implemented in the single and multiple modules, similar to those currently implemented for option listeners and search input listeners, but I think that's acceptable and might even be beneficial for future add-on modules.

I noticed a few minor issues, but I'll add inline comments for those. Also, I would appreciate a unit test for the added functionality.

Thanks a lot!

File: src/selectivity-multiple.js [hunk 1]
```diff
@@ -455,6 +468,8 @@ var callSuper = Selectivity.inherits(MultipleSelectivity, {
      */
     _rerenderSelection: function(event) {
 
+        var $select = this.$('select');
```
Comment (arendjr) [inline src/selectivity-multiple.js]: Please move this statement down to where `$select` is actually used.

File: src/selectivity-single.js [hunk 2]
```diff
@@ -218,19 +224,28 @@ var callSuper = Selectivity.inherits(SingleSelectivity, {
...
+            if ($select.length) {
+                $select.html(this.template('selectOptionCompliance', options));
+            }
         } else {
             $container.html(
                 this.template('singleSelectPlaceholder', { placeholder: this.options.placeholder })
             );
+
+            if ($select.length) {
+                $select.html('');
```
Comment (arendjr) [inline src/selectivity-single.js]: I think using `$select.empty()` is more appropriate.

File: src/selectivity-templates.js [hunk 3]
```diff
@@ -151,6 +151,15 @@ Selectivity.Templates = {
     },
 
     /**
+     * Renders select-box inside single-select input that was initialized on
+     * traditional <select> element.
+     *
+     */
+    multipleSelectCompliance: function(name) {
+      return ('<select name="' + name + '" multiple></select>');
```
Comment (arendjr) [inline src/selectivity-templates.js]: Please use 4 spaces for indentation.

File: src/selectivity-templates.js [hunk 4]
```diff
@@ -228,6 +237,15 @@ Selectivity.Templates = {
     ),
 
     /**
+     * Renders select-box inside single-select input that was initialized on
+     * traditional <select> element.
+     *
+     */
+    singleSelectCompliance: function(name) {
```
Comment (arendjr) [inline src/selectivity-templates.js]: Might be nice idea to use a single template with `multiple` as parameter.

File: src/selectivity-templates.js [hunk 5]
```diff
@@ -267,6 +285,22 @@ Selectivity.Templates = {
...
             '</span>'
         );
+    },
+
+    /**
+     * Renders the selected item in compliance <select> element as <option>.
+     *
+     * @param options Options object containing the following properties
+     *                id - Identifier for the item.
+     *                removable - Boolean whether a remove icon should be displayed.
```
Comment (arendjr) [inline src/selectivity-templates.js]: `removable` can be removed from the docs here.

File: src/selectivity-multiple.js [hunk 6]
```diff
@@ -439,10 +445,17 @@ var callSuper = Selectivity.inherits(MultipleSelectivity, {
...
             highlighted: (item.id === this._highlightedItemId),
             removable: !this.options.readOnly
-        }, item)));
+        }, item);
+
+        this.$searchInput.before(this.template('multipleSelectedItem', options));
+
+        var $select = this.$('select');
+        if ($select.length) {
+            $select.append(this.template('selectOptionCompliance', options));
```
Comment (arendjr) [inline src/selectivity-multiple.js]: I'm wondering whether this code should actually be in `_rerenderSelection()`...

File: src/selectivity-templates.js (hunk 3 above)
Comment (msx2) [inline src/selectivity-templates.js]: Sorry, fixed.

File: src/selectivity-single.js (hunk 2 above)
Comment (msx2) [inline src/selectivity-single.js]: You are definitely right. Done!

File: src/selectivity-multiple.js (hunk 1 above)
Comment (msx2) [inline src/selectivity-multiple.js]: Done.

File: src/selectivity-multiple.js (hunk 6 above)
Comment (msx2) [inline src/selectivity-multiple.js]: Well, this is the place where actual selections are rendered

File: src/selectivity-templates.js (hunk 4 above)
Comment (msx2) [inline src/selectivity-templates.js]: Done, used `selectCompliance`.

File: src/selectivity-templates.js (hunk 5 above)
Comment (msx2) [inline src/selectivity-templates.js]: Deleted. I've kept it as this is the same list of options as the other template uses

**Comment by msx2** [comment]
At first I started digging directly in the traditional module, but I got the feeling it would be an overkill to create additional hooks/events to handle this. Especially that whatever comes from the traditional module is being replaced in single/multiple modules.

If you have an idea in mind of how to refactor this to be completely module-independent, let me know.

**Comment by msx2** [comment]
As for the tests - will something like https://github.com/msx2/selectivity/commit/aa77fea3f83cccc61997d0c0bff534ab95dc601c suffice?

File: dist/selectivity-full.js [hunk 7]
```diff
@@ -4407,6 +4440,31 @@ Selectivity.Templates = {
...
+     *
+     * @param options Options object containing the following properties
+     *                id - Identifier for the item.
+     *                text - Text label which the user sees.
+     */
+    selectOptionCompliance: function(options) {
+        return (
+            '<option value="' + escape(options.id) + '" selected>' +
+              escape(options.text) +
+            '</option'
```
Comment (arendjr) [inline dist/selectivity-full.js]: Looks like a closing `>` is missing.

File: dist/selectivity-full.js [hunk 8]
```diff
@@ -3567,10 +3573,17 @@ var callSuper = Selectivity.inherits(MultipleSelectivity, {
...
     _renderSelectedItem: function(item) {
 
-        this.$searchInput.before(this.template('multipleSelectedItem', $.extend({
+        var options = $.extend({
             highlighted: (item.id === this._highlightedItemId),
             removable: !this.options.readOnly
-        }, item)));
+        }, item);
+
+        this.$searchInput.before(this.template('multipleSelectedItem', options));
```
Comment (arendjr) [inline dist/selectivity-full.js]: I think you can leave this as it was before without assigning the options variable, and just passing `item` to the `selectOptionCompliance` template.

File: src/selectivity-multiple.js (hunk 6 above)
Comment (arendjr) [inline src/selectivity-multiple.js]: Ah yes, I overlooked this is actually called form _rerenderSelection() itself :)

File: tests/unit/traditional.js [hunk 9]
```diff
@@ -30,5 +32,7 @@ exports.testInitializationMultiple = DomUtil.createDomTest(
         test.deepEqual($input.selectivity('value'), [3, 4]);
 
         test.deepEqual($('#selectivity-input').selectivity('value'), [3, 4]);
+
+        test.equal($input.find('select[name="my_select"] option[selected]').length, 2);
```
Comment (arendjr) [inline tests/unit/traditional.js]: This is a good start. To be complete I think you want to test the values of the `<option>` elements.

Comment (msx2) [inline tests/unit/traditional.js]: Um, could you give me a hand here? I tried to use `$input.find('select[name="my_select"]').val()` that normally returns a string value/array of strings, but got `undefined` instead. I tried also `$input.find('select[name="my_select"] option[selected]').first` and `$input.find('select[name="my_select"] option[selected]')[0]` but also got `undefined` in both cases.

**Comment by arendjr** [comment]
I think I might have a solution for the refactoring to get the logic out of the single and multiple modules. Right after _rerenderSelection() is called in the constructors, you can trigger a `selectivity-init` event. This event simply says nothing more than the Selectivity instance has been initialized. The traditional module can now watch the `selectivity-init change` events and update the options in the `<select>` element accordingly. In the case `selectivity-init` you also insert the `<select>` element itself.

I think by using this approach you might even be able to avoid a little duplication between the single and multiple modules as it currently is. The only thing you need to be careful about is that you need to update based on the value, which can be either an array (indicating a multiple-value input) or null or an ID (indicating a single-value input).

**Comment by msx2** [comment]
@arendjr Please take a look now. :smile: 
I must admit I'm not fully happy with event bindings though.

**Comment by arendjr** [comment]
Great! Now the final improvement might be instead of the `if (event.added) { ... } else if (event.removed) { ... }` block, you might want to unconditionally rerender all the selected options. I think that might even solve some edge cases when the changes are not initiated by the user (I think the `_rerenderSelection()` in the multiple module also has a third else-clause to handle that).

What makes you unhappy about using event bindings? I think they're a natural place for add-ons to attach to :)

File: src/selectivity-templates.js [hunk 10]
```diff
@@ -267,6 +267,33 @@ Selectivity.Templates = {
...
+     * Renders the selected item in compliance <select> element as <option>.
+     *
+     * @param options Options object containing the following properties
+     *                id - Identifier for the item.
+     *                text - Text label which the user sees.
+     */
+    selectOptionCompliance: function(options) {
+        return (
+            '<option value="' + escape(options.id) + '" selected>' +
+              escape(options.text) +
```
Comment (arendjr) [inline src/selectivity-templates.js]: Two spaces missing here ;)

**Comment by msx2** [comment]
Do you have in mind rerendering based on the present `.selectivity-multiple-selected-item` elements?

**Comment by arendjr** [comment]
No, just based on the value of `selectivity._data` that you're already checking.

**Comment by msx2** [comment]
Done in https://github.com/msx2/selectivity/commit/11bd36a8ac8e22a6db8c0c3c6ff5cd96fc347615

**Comment by msx2** [comment]
@arendjr I think it might be a good idea to create a new release (e.g. 1.0.1, or 1.1.0) after this PR gets merged. ;)

**Comment by arendjr** [comment]
Okay, gonna merge now. Great work!

I agree about the new release, but I do want to finish the keyboard support improvements I started working on. Hope I can look into that on Sunday or Monday.

Thanks!

**Pull request merged**